### ISSUE TITLE
Enhancement: Tile build progress indicator

### ISSUE THREAD

**Issue by SwedishBarbarossa** [valhalla/valhalla#4801]
I'm currently building the tiles for planet-latest.osm.pbf for a personal project using the gis-ops docker image.

My workstation is running Ubuntu with an **i7-3770 and 8 GB** RAM. Which is definitely on the lower end.

It's taking days, which is expected and fine. Well, I didn't expect it going in, but after looking at what the code does it figures.

**However, some of the slower steps give no progress indication.** Other parts do, which suggests you are open to it. Looking at the code it looks like the slow part is a couple of loops. And I wonder if it is possible to add some progress bar to them? It would aid with expectations and planning.

I am not handy in C++, so I don't think any contributions from me would be helpful.

## Notable places:
[Parsing ways...](https://github.com/valhalla/valhalla/blob/9725261d1c33d6fa361d3d0eec12404f79ca150d/src/mjolnir/pbfgraphparser.cc#L5022)
Console exerpt:
```
valhalla-1  | 2024/07/14 11:36:50.131309 [INFO] Parsing ways...
[truncated most errors]
valhalla-1  | 2024/07/14 16:46:57.123219 [WARN] Error during parsing of `layer` tag on the way 1298158378: stoi
valhalla-1  | 2024/07/14 17:59:35.722526 [INFO] Added 342203 culdesac roundabouts from 567782 candidates.
valhalla-1  | 2024/07/14 17:59:35.730586 [INFO] Finished with 232542595 routable ways containing 2613495508 nodes
```
Total time: 6h 32m

[Reclassifying_V2 link graph edges...](https://github.com/valhalla/valhalla/blob/9725261d1c33d6fa361d3d0eec12404f79ca150d/src/mjolnir/linkclassification.cc#L906)
Console exerpt:
```
valhalla-1  | 2024/07/15 09:32:55.014417 [INFO] Reclassifying_V2 link graph edges...
```

Had to power down the PC at 24/07/16 01:18 for unrelated reasons. That step didn't finish and seems to rebuild from scratch now... Heavily bottle-necked by read/write speed of my HDD.
Total time: over 15 hours

There are a few others which have taken a couple of hours, but just those two would probably help a lot. Dropping my console log as a file if you want to do it in more places.
[log.txt](https://github.com/user-attachments/files/16242193/log.txt)

On a separate note, are there any prebuilt files available? My project doesn't require particularly up to date maps - up to 5 years old is acceptable.

**Comment by kevinkreiser** [comment]
yes for the parsing step we could make a progress bar so long as parsing stays single threaded. we'd very much like for it to become multithreaded because this would be a heck of a lot faster at which point it would make doing progress bars pretty difficult. at any rate we could do the progress bar and throw it away after we do multithreading i suppose.

so for the progress bar we'd want to do a pass over the pbf once to count all the things. this is done pretty easily i think and then the main parser loop can be updated to log once at the beginning about how much there is to parse and again in the loop every so often. would this fill your needs? other than the parser you were saying about reclassify links needing progress updates? i didnt look closely about what would make sense there but its likely possible to do some updates since we have a step to go fill a work queue and then we burn down the work queue.

by the way:

> That step didn't finish and seems to rebuild from scratch now... 

if you mean it started over at the beginning that is a shame. you can tell it to resume where it left off by specifying the stage at which you want it to start. reclassify links happens in the graph build stage so you could have skipped the parsing stage.

**Comment by SwedishBarbarossa** [comment]
> we could do the progress bar and throw it away after we do multithreading i suppose.

Just a thought, I don't know what I'm talking about: Would it be possible to prepare a determinate length queue of tasks, process them in threads and make a callback to iterate the progress when returning?

> so for the progress bar we'd want to do a pass over the pbf once to count all the things.

Curious how long such a pass would take - I have no conception of it. I trust your judgement.

> then the main parser loop can be updated to log once at the beginning about how much there is to parse and again in the loop every so often. would this fill your needs?

I imagine that would look something like this, and would absolutely be sufficient (from the "Building admin db" step):
```
valhalla-1  | 2024/07/14 10:20:14.537878 [INFO] Parsing nodes...
valhalla-1  | 2024/07/14 10:20:29.477243 [INFO] Processed 500000 nodes on ways
valhalla-1  | 2024/07/14 10:20:54.478445 [INFO] Processed 1000000 nodes on ways
valhalla-1  | 2024/07/14 10:21:13.169553 [INFO] Processed 1500000 nodes on ways
valhalla-1  | 2024/07/14 10:21:33.897401 [INFO] Processed 2000000 nodes on ways
valhalla-1  | 2024/07/14 10:21:51.901061 [INFO] Processed 2500000 nodes on ways
valhalla-1  | 2024/07/14 10:22:40.043138 [INFO] Processed 3000000 nodes on ways
valhalla-1  | 2024/07/14 10:23:19.047248 [INFO] Processed 3500000 nodes on ways
valhalla-1  | 2024/07/14 10:23:46.666549 [INFO] Processed 4000000 nodes on ways
valhalla-1  | 2024/07/14 10:24:24.399803 [INFO] Processed 4500000 nodes on ways
valhalla-1  | 2024/07/14 10:24:58.761944 [INFO] Processed 5000000 nodes on ways
valhalla-1  | 2024/07/14 10:25:35.636414 [INFO] Processed 5500000 nodes on ways
valhalla-1  | 2024/07/14 10:26:14.918528 [INFO] Processed 6000000 nodes on ways
valhalla-1  | 2024/07/14 10:27:08.237675 [INFO] Processed 6500000 nodes on ways
valhalla-1  | 2024/07/14 10:27:42.832904 [INFO] Processed 7000000 nodes on ways
valhalla-1  | 2024/07/14 10:28:25.258396 [INFO] Processed 7500000 nodes on ways
valhalla-1  | 2024/07/14 10:28:56.887019 [INFO] Processed 8000000 nodes on ways
valhalla-1  | 2024/07/14 10:30:02.383947 [INFO] Processed 8500000 nodes on ways
valhalla-1  | 2024/07/14 10:30:44.073690 [INFO] Processed 9000000 nodes on ways
valhalla-1  | 2024/07/14 10:31:47.882671 [INFO] Processed 9500000 nodes on ways
valhalla-1  | 2024/07/14 10:32:26.747943 [INFO] Processed 10000000 nodes on ways
valhalla-1  | 2024/07/14 10:33:02.364688 [INFO] Processed 10500000 nodes on ways
valhalla-1  | 2024/07/14 10:33:37.006929 [INFO] Processed 11000000 nodes on ways
valhalla-1  | 2024/07/14 10:34:26.651183 [INFO] Processed 11500000 nodes on ways
valhalla-1  | 2024/07/14 10:35:42.821727 [INFO] Processed 12000000 nodes on ways
valhalla-1  | 2024/07/14 10:36:51.436129 [INFO] Processed 12500000 nodes on ways
valhalla-1  | 2024/07/14 10:37:51.564192 [INFO] Processed 13000000 nodes on ways
valhalla-1  | 2024/07/14 10:38:57.739746 [INFO] Processed 13500000 nodes on ways
valhalla-1  | 2024/07/14 10:40:01.367443 [INFO] Processed 14000000 nodes on ways
valhalla-1  | 2024/07/14 10:40:51.409806 [INFO] Processed 14500000 nodes on ways
valhalla-1  | 2024/07/14 11:05:29.268954 [INFO] Finished with 14774093 nodes
```

However this one doesn't say how many nodes there are until it is finished. Which would make it more helpful as one has an endpoint where one can extrapolate an estimated time to completion.

> you could have skipped the parsing stage.

Big RIP. Oh well, good to know if it happens again!

**Comment by nilsnolde** [comment]
we have some multithreading now, at least for parse ways
